include/kernel.h

```diff
@@ -2056,6 +2057,76 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
 
 /** @} */
 
+#ifdef CONFIG_USERSPACE
+/**
+ * @brief futex structure
+ *
+ * k_futex are tracked as kernel objects and can live in user memory
+ * so any access bypasses the kernel object permission management
+ * mechanism.
+ */
+struct k_futex {
+	atomic_t val;
+};
+
+/**
+ * @defgroup futex_apis FUTEX APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief Initialize a futex.
+ *
+ * This routine initializes a futex object, prior to its first use.
+ *
+ * @param futex Address of the k_futex.
+ *
+ * @return N/A
+ */
+__syscall void k_futex_init(struct k_futex *futex);
+
+/**
+ * @brief Pend the current thread on a futex
+ *
+ * Tests that the supplied futex contains the expected value, and if so,
+ * goes to sleep until some other thread calls k_futex_wake() on it.
+ *
+ * @param futex Address of the futex
+ * @param expected Expected value of the futex, if it is different the caller
+ *		   will not wait on it
+ * @param timeout Waiting period on the futex, in milliseconds, or one of the
+ *		  special values K_NO_WAIT or K_FOREVER
+ * @retval -EACCES Caller does not have read access to futex address
+ * @retval -EAGAIN If the futex value did not match the expected parameter
+ * @retval -EINVAL Futex parameter address not recognized by the kernel
+ * @retval -ETIMEDOUT Thread woke up due to timeout and not a futex wakeup
+ * @retval -EBUSY Caller would have slept, but K_NO_WAIT was passed as a
+ *		  timeout parameter
+ * @retval 0 if the caller went to sleep and was woken up. The caller
+ *	     should check the futex's value on wakeup to determine if it needs
+ *	     to block again
+ */
+__syscall int k_futex_wait(struct k_futex *futex, int expected, s32_t timeout);
+
+/**
+ * @brief Wake a specified number of threads pending on a futex
+ *
+ * Wake up at most num_waiters threads pending on the supplied futex.
+ * No guarantees are made on which threads are woken, scheduling priority
```

Comment:
Actually, since we are using z_unpend_first_thread() to wake up threads, the highest priority thread *will* be woken up. Change the documentation to reflect this.